File: src/utils/textPreprocessing.ts
// Enhanced NLP text preprocessing utilities
export class TextPreprocessor {
  private static stopWords = new Set([
    'the', 'is', 'at', 'which', 'on', 'are', 'as', 'able', 'about', 'above',
    'according', 'accordingly', 'across', 'actually', 'after', 'afterwards',
    'again', 'against', 'aint', 'all', 'allow', 'allows', 'almost', 'alone',
    'along', 'already', 'also', 'although', 'always', 'am', 'among', 'amongst',
    'an', 'and', 'another', 'any', 'anybody', 'anyhow', 'anyone', 'anything',
    'anyway', 'anyways', 'anywhere', 'apart', 'appear', 'appreciate',
    'appropriate', 'around', 'aside', 'ask', 'asking', 'associated', 'available',
    'away', 'awfully', 'be', 'became', 'because', 'become', 'becomes', 'becoming',
    'been', 'before', 'beforehand', 'behind', 'being', 'believe', 'below',
    'beside', 'besides', 'best', 'better', 'between', 'beyond', 'both', 'brief',
    'but', 'by', 'came', 'can', 'cannot', 'cant', 'cause', 'causes', 'certain',
    'certainly', 'changes', 'clearly', 'co', 'com', 'come', 'comes', 'concerning',
    'consequently', 'consider', 'considering', 'contain', 'containing', 'contains',
    'corresponding', 'could', 'course', 'currently', 'said', 'says', 'according',
    'reported', 'reports', 'news', 'article', 'story', 'breaking', 'update', 'latest'
  ]);

  static cleanText(text: string): string {
    if (!text || typeof text !== 'string') return '';
    
    return text
      .toLowerCase()
      // Remove URLs
      .replace(/https?:\/\/[^\s]+/g, ' ')
      // Remove email addresses
      .replace(/\S+@\S+/g, ' ')
      // Remove HTML tags
      .replace(/<[^>]*>/g, ' ')
      // Remove non-alphabetic characters except spaces
      .replace(/[^a-zA-Z\s]/g, ' ')
      // Normalize whitespace
      .replace(/\s+/g, ' ')
      .trim();
  }

  static removeStopWords(text: string): string {
    return text
      .split(' ')
      .filter(word => word.length > 2 && !this.stopWords.has(word))
      .join(' ');
  }

  static stemWord(word: string): string {
    if (!word || word.length < 3) return word;
    
    word = word.toLowerCase();
    
    // Enhanced stemming rules
    const stemmingRules = [
      { suffix: 'ational', replacement: 'ate' },
      { suffix: 'tional', replacement: 'tion' },
      { suffix: 'enci', replacement: 'ence' },
      { suffix: 'anci', replacement: 'ance' },
      { suffix: 'izer', replacement: 'ize' },
      { suffix: 'alli', replacement: 'al' },
      { suffix: 'entli', replacement: 'ent' },
      { suffix: 'eli', replacement: 'e' },
      { suffix: 'ousli', replacement: 'ous' },
      { suffix: 'ization', replacement: 'ize' },
      { suffix: 'ation', replacement: 'ate' },
      { suffix: 'ator', replacement: 'ate' },
      { suffix: 'alism', replacement: 'al' },
      { suffix: 'iveness', replacement: 'ive' },
      { suffix: 'fulness', replacement: 'ful' },
      { suffix: 'ousness', replacement: 'ous' },
      { suffix: 'aliti', replacement: 'al' },
      { suffix: 'iviti', replacement: 'ive' },
      { suffix: 'biliti', replacement: 'ble' },
      { suffix: 'ing', replacement: '' },
      { suffix: 'ly', replacement: '' },
      { suffix: 'ed', replacement: '' },
      { suffix: 'ies', replacement: 'i' },
      { suffix: 'ied', replacement: 'i' },
      { suffix: 'ies', replacement: 'i' },
      { suffix: 's', replacement: '' }
    ];

    for (const rule of stemmingRules) {
      if (word.endsWith(rule.suffix) && word.length > rule.suffix.length + 2) {
        return word.slice(0, -rule.suffix.length) + rule.replacement;
      }
    }
    
    return word;
  }

  static preprocess(text: string): string {
    const cleaned = this.cleanText(text);
    const withoutStopWords = this.removeStopWords(cleaned);
    const stemmed = withoutStopWords
      .split(' ')
      .map(word => this.stemWord(word))
      .filter(word => word.length > 1)
      .join(' ');
    
    return stemmed;
  }

  static extractFeatures(text: string): { [key: string]: number } {
    const features: { [key: string]: number } = {};
    
    if (!text || typeof text !== 'string') {
      return features;
    }
    
    // Basic text statistics
    const words = text.split(/\s+/).filter(word => word.length > 0);
    const sentences = text.split(/[.!?]+/).filter(s => s.trim().length > 0);
    
    features.char_count = text.length;
    features.word_count = words.length;
    features.sentence_count = sentences.length;
    
    // Word-level features
    if (words.length > 0) {
      features.avg_word_length = words.reduce((sum, word) => sum + word.length, 0) / words.length;
      features.long_words = words.filter(word => word.length > 6).length;
      features.short_words = words.filter(word => word.length <= 3).length;
    }
    
    // Punctuation features
    features.exclamation_count = (text.match(/!/g) || []).length;
    features.question_count = (text.match(/\?/g) || []).length;
    features.period_count = (text.match(/\./g) || []).length;
    
    // Ratio features
    if (text.length > 0) {
      features.punctuation_ratio = (text.match(/[.!?,;:]/g) || []).length / text.length;
      features.caps_ratio = (text.match(/[A-Z]/g) || []).length / text.length;
      features.digit_ratio = (text.match(/\d/g) || []).length / text.length;
    }
    
    // Sensational language detection
    const sensationalWords = [
      'shocking', 'amazing', 'incredible', 'unbelievable', 'stunning',
      'miracle', 'secret', 'hidden', 'exposed', 'revealed', 'urgent',
      'breaking', 'exclusive', 'conspiracy', 'scandal', 'forbidden',
      'dangerous', 'terrifying', 'outrageous', 'devastating'
    ];
    
    const textLower = text.toLowerCase();
    features.sensational_count = sensationalWords.filter(word => 
      textLower.includes(word)
    ).length;
    
    features.sensational_ratio = words.length > 0 ? 
      features.sensational_count / words.length : 0;
    
    // Credibility indicators
    const credibilityWords = [
      'study', 'research', 'university', 'scientist', 'doctor',
      'professor', 'evidence', 'data', 'analysis', 'report',
      'journal', 'published', 'peer-reviewed', 'clinical'
    ];
    
    features.credibility_count = credibilityWords.filter(word => 
      textLower.includes(word)
    ).length;
    
    features.credibility_ratio = words.length > 0 ? 
      features.credibility_count / words.length : 0;
    
    // Readability approximation (Flesch Reading Ease)
    if (sentences.length > 0 && words.length > 0) {
      const avgSentenceLength = words.length / sentences.length;
      const avgSyllables = this.estimateAvgSyllables(words);
      features.readability_score = 206.835 - (1.015 * avgSentenceLength) - (84.6 * avgSyllables);
    }
    
    return features;
  }

  private static estimateAvgSyllables(words: string[]): number {
    const totalSyllables = words.reduce((total, word) => {
      return total + this.countSyllables(word);
    }, 0);
    
    return words.length > 0 ? totalSyllables / words.length : 0;
  }

  private static countSyllables(word: string): number {
    if (!word || word.length <= 3) return 1;
    
    word = word.toLowerCase().replace(/[^a-z]/g, '');
    
    // Count vowel groups
    let syllableCount = 0;
    let previousWasVowel = false;
    
    for (let i = 0; i < word.length; i++) {
      const isVowel = 'aeiouy'.includes(word[i]);
      
      if (isVowel && !previousWasVowel) {
        syllableCount++;
      }
      
      previousWasVowel = isVowel;
    }
    
    // Handle silent 'e'
    if (word.endsWith('e')) {
      syllableCount--;
    }
    
    // Minimum one syllable
    return Math.max(1, syllableCount);
  }
}

File: backend/text_preprocessing.py
import re
import string
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from nltk.tokenize import word_tokenize
import nltk

# Download required NLTK data
try:
    nltk.data.find('tokenizers/punkt')
except LookupError:
    nltk.download('punkt')

try:
    nltk.data.find('corpora/stopwords')
except LookupError:
    nltk.download('stopwords')

class TextPreprocessor:
    def __init__(self):
        self.stemmer = PorterStemmer()
        self.stop_words = set(stopwords.words('english'))
        
        # Add custom stop words for news domain
        custom_stops = {
            'said', 'says', 'according', 'reported', 'reports', 'news',
            'article', 'story', 'breaking', 'update', 'latest', 'new'
        }
        self.stop_words.update(custom_stops)
    
    def clean_text(self, text):
        """Clean and normalize text"""
        if not isinstance(text, str):
            return ""
        
        # Convert to lowercase
        text = text.lower()
        
        # Remove URLs
        text = re.sub(r'http\S+|www\S+|https\S+', '', text, flags=re.MULTILINE)
        
        # Remove email addresses
        text = re.sub(r'\S+@\S+', '', text)
        
        # Remove HTML tags
        text = re.sub(r'<.*?>', '', text)
        
        # Remove special characters and digits, keep only letters and spaces
        text = re.sub(r'[^a-zA-Z\s]', ' ', text)
        
        # Remove extra whitespace
        text = re.sub(r'\s+', ' ', text).strip()
        
        return text
    
    def tokenize_text(self, text):
        """Tokenize text into words"""
        try:
            tokens = word_tokenize(text)
            return tokens
        except:
            # Fallback tokenization
            return text.split()
    
    def remove_stopwords(self, tokens):
        """Remove stop words from tokens"""
        return [token for token in tokens if token not in self.stop_words and len(token) > 2]
    
    def stem_tokens(self, tokens):
        """Apply stemming to tokens"""
        return [self.stemmer.stem(token) for token in tokens]
    
    def preprocess(self, text):
        """Complete preprocessing pipeline"""
        if not text or not isinstance(text, str):
            return ""
        
        # Clean text
        cleaned_text = self.clean_text(text)
        
        if not cleaned_text:
            return ""
        
        # Tokenize
        tokens = self.tokenize_text(cleaned_text)
        
        # Remove stop words
        filtered_tokens = self.remove_stopwords(tokens)
        
        # Apply stemming
        stemmed_tokens = self.stem_tokens(filtered_tokens)
        
        # Join back into string
        processed_text = ' '.join(stemmed_tokens)
        
        return processed_text
    
    def extract_features(self, text):
        """Extract various text features for analysis"""
        if not text or not isinstance(text, str):
            return {}
        
        features = {}
        
        # Basic text statistics
        features['char_count'] = len(text)
        features['word_count'] = len(text.split())
        features['sentence_count'] = len(re.split(r'[.!?]+', text))
        
        # Punctuation features
        features['exclamation_count'] = text.count('!')
        features['question_count'] = text.count('?')
        features['punctuation_ratio'] = sum(1 for c in text if c in string.punctuation) / len(text) if text else 0
        
        # Capitalization features
        features['caps_count'] = sum(1 for c in text if c.isupper())
        features['caps_ratio'] = features['caps_count'] / len(text) if text else 0
        
        # Word-level features
        words = text.split()
        if words:
            features['avg_word_length'] = sum(len(word) for word in words) / len(words)
            features['long_words'] = sum(1 for word in words if len(word) > 6)
            features['short_words'] = sum(1 for word in words if len(word) <= 3)
        else:
            features['avg_word_length'] = 0
            features['long_words'] = 0
            features['short_words'] = 0
        
        # Sensational language indicators
        sensational_words = [
            'shocking', 'amazing', 'incredible', 'unbelievable', 'stunning',
            'miracle', 'secret', 'hidden', 'exposed', 'revealed', 'urgent',
            'breaking', 'exclusive', 'conspiracy', 'scandal'
        ]
        
        text_lower = text.lower()
        features['sensational_count'] = sum(1 for word in sensational_words if word in text_lower)
        features['sensational_ratio'] = features['sensational_count'] / len(words) if words else 0
        
        return features
    
    def get_readability_score(self, text):
        """Calculate simple readability metrics"""
        if not text or not isinstance(text, str):
            return 0
        
        words = text.split()
        sentences = re.split(r'[.!?]+', text)
        sentences = [s.strip() for s in sentences if s.strip()]
        
        if not words or not sentences:
            return 0
        
        # Simple readability approximation
        avg_sentence_length = len(words) / len(sentences)
        avg_word_length = sum(len(word) for word in words) / len(words)
        
        # Lower scores indicate easier readability
        readability = (avg_sentence_length * 1.015) + (avg_word_length * 84.6) - 206.835
        
        return max(0, min(100, readability))